src/Typst/Parse.hs

```diff
@@ -433,44 +439,71 @@ mSymbol =
   lexeme ( Text . T.singleton
             <$> satisfy (\c -> not (isSpace c) && c /= '$' && c /= '\\'))
 
-withIndent :: Int -> P a -> P a
-withIndent indent pa = do
-  oldIndent <- stIndent <$> getState
-  updateState $ \st -> st {stIndent = indent : oldIndent}
-  ms <- pa
-  updateState $ \st -> st {stIndent = oldIndent}
-  pure ms
+-- Parser that discards leading trivia, and does not parse trailing trivia
+-- unless followed by non-trivia markup. Stops when non-trivia markup would
+-- match `stop` parser.
+-- Does not allow new lines except in block comments.
+pSinglelineWhileNot :: P a -> P [Markup]
+pSinglelineWhileNot stop = do
+  option [] $ (concat . drop 1 . concat) <$> many1 pSinglelineChunk
+  where
+    pSinglelineChunk :: P [[Markup]]
+    pSinglelineChunk = try $ do
+      trivia <- many1 pTriviaWithoutEol
+      markup <- manyNonTriviaMarkupExcept
+      guard (not (null markup))
+      pure [trivia, markup]
+      where
+        manyNonTriviaMarkupExcept = (lookAhead (try stop) >> pure [])
+                 <|> (pNonTriviaMarkup >>= \x -> (x:) <$> manyNonTriviaMarkupExcept)
+                 <|> pure []   
```

Comment:
Isn't this the same as `many (notFollowedBy (try stop) >> pNonTrivaMarkup)`?
which seems easier to understand

Reply:
For better performance (I think) one could omit the `try` before `stop` and just ensure that the parsers passed in as `stop` have a `try` if they need it.  Many of them, e.g. `char ':'`, don't need a `try` and the `try` will just slow things down.